README.md
# metasdk-react

[![License](http://img.shields.io/badge/license-MIT-blue.svg)](https://raw.githubusercontent.com/hexoul/metasdk-react/master/LICENSE)
[![NPM](https://img.shields.io/npm/v/metasdk-react.svg)](https://www.npmjs.com/package/metasdk-react)
[![JavaScript Style Guide](https://img.shields.io/badge/code_style-standard-brightgreen.svg)](https://standardjs.com)

> SDK to communicate between React and App using QRCode

## Install

```bash
npm i metasdk-react
```

## Components

1. Login
2. Request
2. SendTransaction

## Usage

If `callbackUrl` will be given as prop, a component shows QR code directly, not popup button.

```jsx
import React, { Component } from 'react'
import { Login, Request, SendTransaction } from 'metasdk-react'

class Example extends Component {
  callbackExample(arg) {}
  
  render () {
    return (
      <div>
        <Login
          data='testmsg'
          service='example'
          qrpopup={true}
          callback={this.callbackExample}
        />

        <Request
          request={['10', '20', '30']}
          usage='example'
          qrpopup={true}
          callback={this.callbackExample}
        />

        <SendTransaction
          id='sendTransactionByRequest'
          request={this.trxRequest}
          usage='method'
          qrpopup={true}
          callback={this.callbackExample}
        />

        <SendTransaction
          id='sendTransactionWithCallbackURL'
          request={this.trxRequest}
          usage='method'
          qrpopup={false}
          callbackUrl='http://localhost/callback'
        />

        <SendTransaction
          id='sendTransactionByRaw'
          to='0x8101487270f5411cf213b8d348a2ab46df66245d'
          value='0x01'
          data='0x02'
          qrsize={256}
          qrvoffset={20}
          qrpadding='2em'
          qrposition='bottom right'
          qrtext='SendTransaction'
          qrpopup={true}
          callback={this.callbackExample}
        />
      </div>
    )
  }
}
```

QRCode styles can be set for all components like above example `SendTransaction`
- `qrpopup` decides if QRCode exists within popup or not. default: `false`
- `qrsize` changes the size of QRCode by pixel. default: `128`
- `qrvoffset` is a vertical offset from origin. default: `20`
- `qrpadding` applies padding thickness of QRCode as style. default: `1em`
- `qrposition` decides relative position from origin, declaration for positions are in `util.js`. default: `bottom right`
- `qrtext` sets `Button` text to open QRCode popup

## Test

```
cd metasdk-react
npm start
[Ctrl+C]
cd example
npm start
```


CHANGELOG.md
# Change Log

All notable changes to this project will be documented in this file.
See [Conventional Commits](https://conventionalcommits.org) for commit guidelines.

## [0.6.0](https://github.com/hexoul/metasdk-react/compare/v0.6.0...v0.5.4) (2018-12-28)


### Bug Fixes

* **config:** Fix rollup config about babel for async jobs ([9275ba0](https://github.com/hexoul/metasdk-react/commit/9275ba0))


### Features

* **Login, Request, SendTransaction:**
  * Make QR code with logo ([941d518](https://github.com/hexoul/metasdk-react/commit/941d518))
  * Replace qrcode.react with qrcode.es and update README ([1e9bab7](https://github.com/hexoul/metasdk-react/commit/1e9bab7)) ([86599eb](https://github.com/hexoul/metasdk-react/commit/86599eb))

* **util:** Refactoring for qrcode.es ([4c9a7e3](https://github.com/hexoul/metasdk-react/commit/4c9a7e3)) ([1631287](https://github.com/hexoul/metasdk-react/commit/1631287))


example/src/App.js
import React, { Component } from 'react'
import { Login, Request, SendTransaction } from 'metasdk-react'
import Web3 from 'web3'

const compiledBNB = [{ 'constant': true, 'inputs': [], 'name': 'name', 'outputs': [{ 'name': '', 'type': 'string' }], 'payable': false, 'type': 'function' }, { 'constant': false, 'inputs': [{ 'name': '_spender', 'type': 'address' }, { 'name': '_value', 'type': 'uint256' }], 'name': 'approve', 'outputs': [{ 'name': 'success', 'type': 'bool' }], 'payable': false, 'type': 'function' }, { 'constant': true, 'inputs': [], 'name': 'totalSupply', 'outputs': [{ 'name': '', 'type': 'uint256' }], 'payable': false, 'type': 'function' }, { 'constant': false, 'inputs': [{ 'name': '_from', 'type': 'address' }, { 'name': '_to', 'type': 'address' }, { 'name': '_value', 'type': 'uint256' }], 'name': 'transferFrom', 'outputs': [{ 'name': 'success', 'type': 'bool' }], 'payable': false, 'type': 'function' }, { 'constant': true, 'inputs': [], 'name': 'decimals', 'outputs': [{ 'name': '', 'type': 'uint8' }], 'payable': false, 'type': 'function' }, { 'constant': false, 'inputs': [{ 'name': 'amount', 'type': 'uint256' }], 'name': 'withdrawEther', 'outputs': [], 'payable': false, 'type': 'function' }, { 'constant': false, 'inputs': [{ 'name': '_value', 'type': 'uint256' }], 'name': 'burn', 'outputs': [{ 'name': 'success', 'type': 'bool' }], 'payable': false, 'type': 'function' }, { 'constant': false, 'inputs': [{ 'name': '_value', 'type': 'uint256' }], 'name': 'unfreeze', 'outputs': [{ 'name': 'success', 'type': 'bool' }], 'payable': false, 'type': 'function' }, { 'constant': true, 'inputs': [{ 'name': '', 'type': 'address' }], 'name': 'balanceOf', 'outputs': [{ 'name': '', 'type': 'uint256' }], 'payable': false, 'type': 'function' }, { 'constant': true, 'inputs': [], 'name': 'owner', 'outputs': [{ 'name': '', 'type': 'address' }], 'payable': false, 'type': 'function' }, { 'constant': true, 'inputs': [], 'name': 'symbol', 'outputs': [{ 'name': '', 'type': 'string' }], 'payable': false, 'type': 'function' }, { 'constant': false, 'inputs': [{ 'name': '_to', 'type': 'address' }, { 'name': '_value', 'type': 'uint256' }], 'name': 'transfer', 'outputs': [], 'payable': false, 'type': 'function' }, { 'constant': true, 'inputs': [{ 'name': '', 'type': 'address' }], 'name': 'freezeOf', 'outputs': [{ 'name': '', 'type': 'uint256' }], 'payable': false, 'type': 'function' }, { 'constant': false, 'inputs': [{ 'name': '_value', 'type': 'uint256' }], 'name': 'freeze', 'outputs': [{ 'name': 'success', 'type': 'bool' }], 'payable': false, 'type': 'function' }, { 'constant': true, 'inputs': [{ 'name': '', 'type': 'address' }, { 'name': '', 'type': 'address' }], 'name': 'allowance', 'outputs': [{ 'name': '', 'type': 'uint256' }], 'payable': false, 'type': 'function' }, { 'inputs': [{ 'name': 'initialSupply', 'type': 'uint256' }, { 'name': 'tokenName', 'type': 'string' }, { 'name': 'decimalUnits', 'type': 'uint8' }, { 'name': 'tokenSymbol', 'type': 'string' }], 'payable': false, 'type': 'constructor' }, { 'payable': true, 'type': 'fallback' }, { 'anonymous': false, 'inputs': [{ 'indexed': true, 'name': 'from', 'type': 'address' }, { 'indexed': true, 'name': 'to', 'type': 'address' }, { 'indexed': false, 'name': 'value', 'type': 'uint256' }], 'name': 'Transfer', 'type': 'event' }, { 'anonymous': false, 'inputs': [{ 'indexed': true, 'name': 'from', 'type': 'address' }, { 'indexed': false, 'name': 'value', 'type': 'uint256' }], 'name': 'Burn', 'type': 'event' }, { 'anonymous': false, 'inputs': [{ 'indexed': true, 'name': 'from', 'type': 'address' }, { 'indexed': false, 'name': 'value', 'type': 'uint256' }], 'name': 'Freeze', 'type': 'event' }, { 'anonymous': false, 'inputs': [{ 'indexed': true, 'name': 'from', 'type': 'address' }, { 'indexed': false, 'name': 'value', 'type': 'uint256' }], 'name': 'Unfreeze', 'type': 'event' }]

export default class App extends Component {
  // Topic number
  // - 10: name
  // - 20: phone
  // - 30: email
  request = ['10', '20', '30']

  componentWillMount () {
    if (typeof window !== 'undefined' && typeof window.web3 !== 'undefined') {
      this.web3 = new Web3(window.web3.currentProvider)
      let BNB = new this.web3.eth.Contract(compiledBNB, '0xdd755ca63bd19299b83ae0bb9fd5bcd765fe908c')
      this.trxRequest = BNB.methods
        .transfer('0x31a46cd5FF7d0DC32b93149dB1A695b68Fc1aa00', this.web3.utils.toWei('0', 'ether'))
        .send.request({ from: '', value: this.web3.utils.toWei('0', 'ether'), gasPrice: '1' })
    }
  }

  callbackExample (arg) {
    console.log('callbackExample', arg)
  }

  requestCallbackExample (arg) {
    console.log('requestCallbackExample', arg)
    this.request.map((req) => {
      console.log('got', req, arg[req])
      return req
    })
  }

  onClickTest () {
    var componentID = 'loginID'
    document.getElementById(componentID).click()
  }

  render () {
    return (
      <div>
        <button onClick={this.onClickTest}>onClickTest</button>

        <Login
          id='loginID'
          data='testmsg'
          service='example'
          qrpopup={true}
          callback={this.callbackExample}
        />
        <hr />

        <Request
          id='requestID'
          request={this.request}
          usage='example'
          metaID='0x31a46cd5FF7d0DC32b93149dB1A695b68Fc1aa00'
          qrsize={256}
          qrpopup={true}
          callback={this.requestCallbackExample}
        />
        <hr />

        <SendTransaction
          id='sendTransactionId1'
          request={this.trxRequest}
          usage='method'
          qrpopup={true}
          callback={this.callbackExample}
        />
        <hr />

        <SendTransaction
          id='sendTransactionWithCallbackURL'
          request={this.trxRequest}
          usage='method'
          qrpopup={false}
          callbackUrl='http://localhost/callback'
        />
        <hr />

        <SendTransaction
          id='sendTransactionId2'
          to='0x8101487270f5411cf213b8d348a2ab46df66245d'
          value={this.web3.utils.toWei('0.01', 'ether')}
          data='data2'
          usage='method'
          qrsize={256}
          qrvoffset={20}
          qrpadding='2em'
          qrposition='bottom right'
          qrtext='SendTransaction'
          qrpopup={true}
          callback={this.callbackExample}
        />
      </div>
    )
  }
}
